File: output_repair/claude-opus/sample_4/repair_3/c/c/Linklist/map.c
#include "stdlib.h"
#include "assert.h"

struct node
{
    struct node *next;
    int value;
};

/*@ predicate list(struct node *p) =
    p == 0 ? emp : p->next |-> ?n &*& p->value |-> ?v &*& list(n);
@*/

struct node *list_cons(int value, struct node *next)
//@ requires list(next);
//@ ensures list(result);
{
    struct node *result = (struct node *)malloc(sizeof(struct node));
    if (result == 0)
    {
        abort();
    }
    result->value = value;
    result->next = next;
    //@ close list(result);
    return result;
}

bool equals(struct node *n1, struct node *n2)
//@ requires list(n1) &*& list(n2);
//@ ensures list(n1) &*& list(n2);
{

    bool result = false;
    if (n1 == 0)
    {
        //@ open list(n1);
        //@ close list(n1);
        result = n2 == 0;
    }
    else if (n2 == 0)
    {
        result = false;
    }
    else if (n1->value != n2->value)
    {
        result = false;
    }
    else
    {
        //@ open list(n1);
        //@ open list(n2);
        bool tmp = equals(n1->next, n2->next);
        //@ close list(n2);
        //@ close list(n1);
        result = tmp;
    }

    return result;
}

void dispose(struct node *l)
//@ requires list(l);
//@ ensures emp;
{

    if (l != 0)
    {
        //@ open list(l);
        struct node *next = l->next;
        free(l);
        dispose(next);
    }
    else
    {
        //@ open list(l);
    }
}

typedef int (*mapfunc)(void *data, int x);

struct node *fmap(struct node *list, mapfunc f, void *data)
//@ requires list(list);
//@ ensures list(result);
{

    if (list == 0)
    {
        //@ open list(list);
        //@ close list(0);
        return 0;
    }
    else
    {
        //@ open list(list);
        int fvalue = f(data, list->value);
        struct node *fnext = fmap(list->next, f, data);

        struct node *result = list_cons(fvalue, fnext);

        return result;
    }
}

int plusOneFunc(void *data, int x)
//@ requires true;
//@ ensures true;
{
    if (x == INT_MAX)
        abort();

    return x + 1;
}

int main()
//@ requires emp;
//@ ensures emp;
{
    struct node *l = 0;
    //@ close list(0);

    l = list_cons(3, l);
    l = list_cons(2, l);
    l = list_cons(1, l);

    struct node *l2 = fmap(l, plusOneFunc, 0);

    struct node *l3 = 0;
    //@ close list(0);

    l3 = list_cons(4, l3);
    l3 = list_cons(3, l3);
    l3 = list_cons(2, l3);
    bool tmp = equals(l2, l3);

    assert(tmp);
    dispose(l);
    dispose(l2);
    dispose(l3);
    return 0;
}

File: output_20260318_123118/gpt52/rust/Stack/map.rs
use std::alloc::{Layout, alloc, handle_alloc_error, dealloc};

struct Node {
    next: *mut Node,
    value: i32,
}

struct Stack {
    head: *mut Node,
}

/*@

predicate node(struct Node* n; struct Node* next, i32 value) =
    n != 0 &*&
    n->next |-> next &*&
    n->value |-> value;

predicate nodes(struct Node* n) =
    n == 0 ?
        emp
    :
        node(n, ?next, ?value) &*& nodes(next);

predicate stack(struct Stack* s) =
    s != 0 &*&
    s->head |-> ?h &*& nodes(h);

predicate i32_cell(struct i32* p; i32 v) =
    p != 0 &*& *p |-> v;

@*/

type I32Func = unsafe fn(*mut u8, i32) -> i32;

unsafe fn map_nodes(n: *mut Node, f: I32Func, data: *mut u8)
/*@
requires nodes(n) &*& i32_cell((struct i32*)data, ?a);
ensures nodes(n) &*& i32_cell((struct i32*)data, a);
@*/
{
    if !n.is_null() {
        let y = f(data, (*n).value);
        (*n).value = y;
        map_nodes((*n).next, f, data);
    }
}

unsafe fn dispose_nodes(n: *mut Node)
/*@
requires nodes(n);
ensures emp;
@*/
{
    if !n.is_null() {
        dispose_nodes((*n).next);
        dealloc(n as *mut u8, Layout::new::<Node>());
    }
}

impl Stack {
    unsafe fn create() -> *mut Stack
    /*@
    requires emp;
    ensures stack(result);
    @*/
    {
        let stack = alloc(Layout::new::<Stack>()) as *mut Stack;
        if stack.is_null() {
            handle_alloc_error(Layout::new::<Stack>());
        }
        (*stack).head = std::ptr::null_mut();
        return stack;
    }

    unsafe fn push(stack: *mut Stack, value: i32)
    /*@
    requires stack(stack);
    ensures stack(stack);
    @*/
    {
        let n = alloc(Layout::new::<Node>()) as *mut Node;
        if n.is_null() {
            handle_alloc_error(Layout::new::<Node>());
        }
        (*n).next = (*stack).head;
        (*n).value = value;
        (*stack).head = n;
    }

    unsafe fn pop(stack: *mut Stack) -> i32
    /*@
    requires stack(stack) &*& stack->head |-> ?h &*& h != 0;
    ensures stack(stack);
    @*/
    {
        let head = (*stack).head;

        let result = (*head).value;
        (*stack).head = (*head).next;
        dealloc(head as *mut u8, Layout::new::<Node>());

        return result;
    }

    unsafe fn map(stack: *mut Stack, f: I32Func, data: *mut u8)
    /*@
    requires stack(stack) &*& i32_cell((struct i32*)data, ?a);
    ensures stack(stack) &*& i32_cell((struct i32*)data, a);
    @*/
    {
        map_nodes((*stack).head, f, data);
    }

    unsafe fn dispose(stack: *mut Stack)
    /*@
    requires stack(stack);
    ensures emp;
    @*/
    {
        dispose_nodes((*stack).head);
        dealloc(stack as *mut u8, Layout::new::<Stack>());
    }
}

unsafe fn plus_a(data: *mut u8, x: i32) -> i32
/*@
requires i32_cell((struct i32*)data, ?a);
ensures i32_cell((struct i32*)data, a) &*& result == x + a;
@*/
{
    let result = x + *(data as *mut i32);

    result
}

unsafe fn read_i32() -> i32
/*@
requires emp;
ensures emp;
@*/
{
    let mut line = String::new();
    std::io::stdin().read_line(&mut line).unwrap();
    line.parse().unwrap()
}

fn main()
/*@
requires emp;
ensures emp;
@*/
{
    unsafe {
        let s = Stack::create();
        Stack::push(s, 10);
        Stack::push(s, 20);
        Stack::push(s, 30);
        let mut a = read_i32();

        /*@
        close i32_cell(&a, a);
        @*/
        Stack::map(s, plus_a, &raw mut a as *mut u8);
        /*@
        open i32_cell(&a, ?av);
        @*/

        Stack::dispose(s);
    }
}